### ISSUE TITLE
sending audio buffer file to node-red using mqtt

### ISSUE THREAD

**Issue by hazee007** [knolleary/pubsubclient#544]
Make your question, not a Statement, inclusive. Include all pertinent information:
hello guys am trying to send audio buffer to node red using the pubsubclient on esp32.
the problem is i am trying to save the audio buffer of 1kb into a char buffer since its the only way i can send data to node-red but i when i gave a char buff[1000], i get no data coming to node-red execpt i set the char buff[110] which is not enough for the voice sample i need.

components am using is esp32 with SparkFun Electret Microphone Breakout(not worried about sound quality)

here the code below

`#include <PubSubClient.h>
#include <WiFi.h>
#include <WiFiClient.h>

const char* ssid = "";
const char* password = "";

const char* mqttServer = "";
const int mqttPort = 1883;

String str_payload;

WiFiClient espClient;
PubSubClient client(espClient);

void setup()
{
Serial.begin(9600);
WiFi.begin(ssid,password);
while(WiFi.status() != WL_CONNECTED) {
delay(500);
Serial.println("Connecting to WiFi..");
}
Serial.println("Connected to the WiFi network");
client.setServer(mqttServer, mqttPort);
while(!client.connected()) {
Serial.println("Connecting to MQTT...");
if(client.connect("ESP8266Client" )) {
Serial.println("connected");
}else{
Serial.print("failed state ");
Serial.print(client.state());
delay(2000);
}
}
vibrate();
my_analog();
}

void loop()
{
}

void vibrate(){
pinMode(2, OUTPUT);
digitalWrite(2,HIGH);
delay(2000);
digitalWrite(2,LOW);
my_record();
}

void my_record()
{
for(int y=0;y<11;y++)
{
for(int i=0;i<1000;i++)
{
str_payload += char(analogRead(35)/4);
// Serial.println(str_payload.length());
}
Send();
}
}

void my_analog()
{
if(digitalRead(2)==LOW)
{
my_record();

}
}

void Send()
{
char buf[110];
String (str_payload).toCharArray(buf,110);
Serial.print("publishing");
client.publish("test/esp8266",buf);
str_payload="";
Serial.println("finish");
}`

**Comment by GitGhidorah** [comment]
Hello.
I think that you have to consider about defined values, (perhaps) when you send more than 1,000 bytes data. 
for example, 
MQTT_MAX_PACKET_SIZE, MQTT_MAX_TRANSFER_SIZE.
Have you checked them?
Sorry,but I don't ecxactly understand the effect of defining MQTT_MAX_TRANSFER_SIZE.

**Comment by hazee007** [comment]
@GitGhidorah yes i did that and was able to send the data to mqtt by found a new problem i wasnt able to get proper voice from the data send most likely due to the fact that am sending the string as a char.
I was able to send the string on esp8266 using an old pubsubclient library but that library isn't working on esp32. ill attach the library below if u can help me in anyway possible.
thanks
[pubsubclientesp8266.zip](https://github.com/knolleary/pubsubclient/files/2716952/pubsubclientesp8266.zip)

**Comment by GitGhidorah** [comment]
Hello.
Mostly,I am a beginner about mqtt and arduino.
So tell me specifically.

> 'i did that'.

  --> What did you do?

>  'the fact that am sending the string as a char.'  

   -->What does it mean?

>  library isn't working on esp32 

   --> What happend?

**Comment by hazee007** [comment]
alright i changed the buffer size of char(MQTT_MAX_PACKET_SIZE, MQTT_MAX_TRANSFER_SIZE.) in the pubsubclient library.

the code is working i.e sending the converted string into char then send to mqtt but the quality of the the received char is bad that i can get any good data from the char sent.

i was able to send the string(not needing to convert to char) on esp8266  using an old library(shared in previous comments)which gives a good output at the mqtt but the library is not working on esp32.

**Comment by GitGhidorah** [comment]
Thank you for your answer.
But I'm sorrry that I can't tell you solutions of this problems.
I don't know the differences between char and String class.
I know only that 'C' string ends with   \x00  .
